File: leetcode/pareto-problem-set/trees/levelOrder.py
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        def bfs(root: Optional[TreeNode]) -> List[List[int]]:
            if not root:
                return []

            queue = [root]
            result = []

            while queue:
                curr_level = []
                sub_arr = []
                start = 0
                while start < len(queue):
                    curr_level.append(queue[start])
                    start += 1

                for node in curr_level:
                    if node.val is not None:
                        sub_arr.append(node.val)
                    if node.left:
                        queue.append(node.left)
                    if node.right:
                        queue.append(node.right)
                result.append(sub_arr)

            return result

        return bfs(root)

File: leetcode/pareto-problem-set/trees/levelOrder.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

import TreeNode from "../../70-blind-leetcode/tree/TreeNode";

function levelOrder(root: TreeNode | null): number[][] {
  function bfs(root: TreeNode | null) {
    if (!root) return [];

    const queue: TreeNode[] = [root];
    const result: number[][] = [];

    while (queue.length > 0) {
      let currLevel: TreeNode[] = [];
      let subArr: number[] = [];
      let start = 0;
      while (start < queue.length) currLevel.push(queue[start++]);

      for (let i = 0; i < currLevel.length; i++) {
        const node = currLevel[i];
        if (node.val !== undefined) {
          subArr.push(node.val);
        }
        if (node.left) queue.push(node.left);
        if (node.right) queue.push(node.right);
      }
      result.push(subArr);
    }

    return result;
  }

  return bfs(root);
}

// Create a binary tree
const root = new TreeNode(3);
root.left = new TreeNode(9);
root.right = new TreeNode(20);
root.right.left = new TreeNode(15);
root.right.right = new TreeNode(7);

// Call the levelOrder function
const result = levelOrder(root);

// Print the result
console.log(result);
